smk/gene_abundance.smk

```diff
@@ -485,19 +488,19 @@ msamtools profile aligned.bam $common_args -o {output.rel_prop_genome} --multi=p
 samtools sort aligned.bam -o sorted.bam -@ {params.sort_threads} -m ${{sort_memory}}G --output-fmt=BAM
 __EOM__
 
-            # Use bedtools multicov to generate read-counts per gene
+            # Use samtools bedcov to generate read-counts per gene
             # 1. Index sorted.bam file
-            # 2. Enable parallelization by splitting the BED file into smaller files and then giving it to bedtools
-            #    a. Create smaller bed files with 10000 lines each, like x0000000.bedsub, x0000001.bedsub, ...
+            # 2. Enable parallelization by splitting the BED file into smaller files and then giving it to samtools bedcov
+            #    a. Create smaller bed files with {params.bedcov_lines} lines each, like x0000000.bedsub, x0000001.bedsub, ...
             #    b. Send 'ls' output to GNU parallel to process these smaller BED files.
 
             parallel --jobs {threads} <<__EOM__
 samtools index sorted.bam sorted.bam.bai -@ {threads}
-split -d --suffix-length=7 --additional-suffix=.bedsub --lines=10000 {input.bed_mini}
+split -d --suffix-length=7 --additional-suffix=.bedsub --lines={params.bedcov_lines} {input.bed_mini}
 __EOM__
 
             # Pipe it to parallel
-            time (ls | grep -E '\.bedsub$' | parallel --jobs {threads} --plus "bedtools multicov -bams sorted.bam -bed {{}} | cut -f4- | (grep -v $'\\t0$' || true) > {{.}}.bed.part")
+            time (ls | grep -E '\.bedsub$' | parallel --jobs {params.bedcov_threads} --plus "samtools bedcov -c -g SECONDARY {{}} sorted.bam | cut -f4,5,7 | (grep -v $'\\t0$' || true) > {{.}}.bed.part")
             (echo -e 'gene_length\\tID\\t{wildcards.omics}.{params.sample_alias}'; cat *.bed.part) | gzip -c > out.bed.gz
```

Comment:
Apart from not wanting to change the final abundances, is there a reason for including the secondary alignments in abundance estimations?

Reply:
In this special case where alignments have been passed through `--besthit` filter from `msamtools`, all alignments for a read are equally good. But some are flagged secondary, because aligners *have to* flag one of them as primary - which is more an artefact of human genome mapping. There is no reason to consider secondary alignments as second-class here.
Also, this was the behavior of `bedtools multicov` that we used previously. This setting for `samtools bedcov` reproduces it exactly.